vm-virtio/src/block.rs

```diff
@@ -94,42 +96,311 @@ impl ExecuteError {
+    }
+
+    pub fn metadata(&self) -> std::io::Result<Metadata> {
+        self.file.metadata()
+    }
+
+    pub fn try_clone(&self) -> std::io::Result<RawFile> {
+        Ok(RawFile {
+            file: self.file.try_clone().expect("RawFile cloning failed"),
+            alignment: self.alignment,
+            position: self.position,
+        })
+    }
+
+    pub fn sync_all(&self) -> std::io::Result<()> {
+        self.file.sync_all()
+    }
+
+    pub fn sync_data(&self) -> std::io::Result<()> {
+        self.file.sync_data()
     }
 }
 
 impl Read for RawFile {
     fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
-        self.file.read(buf)
-    }
-}
+        if self.is_aligned(buf) {
+            let ret = unsafe {
+                ::libc::read(
+                    self.file.as_raw_fd(),
+                    buf.as_mut_ptr() as *mut c_void,
+                    buf.len(),
+                )
+            };
+            if ret < 0 {
+                Err(io::Error::last_os_error())
+            } else {
+                self.position = self.position.checked_add(ret.try_into().unwrap()).unwrap();
+                Ok(ret as usize)
+            }
+        } else {
+            let rounded_pos = self.round_down(self.position);
+            let file_offset = self.position.checked_sub(rounded_pos).unwrap();
+            let buf_len: u64 = buf.len().try_into().unwrap();
+            let rounded_len: usize = self
+                .round_up(file_offset.checked_add(buf_len).unwrap())
+                .try_into()
+                .unwrap();
 
-impl Seek for RawFile {
-    fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
-        self.file.seek(pos)
+            let layout =
+                Layout::from_size_align(rounded_len, self.alignment.try_into().unwrap()).unwrap();
+            let ptr = unsafe { alloc(layout) };
+
+            let ret = unsafe {
+                ::libc::pread(
+                    self.file.as_raw_fd(),
+                    ptr as *mut c_void,
+                    rounded_len,
+                    rounded_pos.try_into().unwrap(),
+                )
+            };
+            if ret < 0 {
+                unsafe { dealloc(ptr, layout) };
+                return Err(io::Error::last_os_error());
+            }
+
+            let read: u64 = ret.try_into().unwrap();
+            if read < file_offset {
+                unsafe { dealloc(ptr, layout) };
+                return Err(io::Error::new(io::ErrorKind::Other, "short read"));
+            }
+
+            let mut to_copy = read - file_offset;
+            if to_copy > buf_len {
+                to_copy = buf_len;
+            }
+
+            unsafe {
+                ::libc::memcpy(
+                    buf.as_mut_ptr() as *mut c_void,
+                    (ptr as u64 + file_offset) as *const c_void,
+                    to_copy.try_into().unwrap(),
+                );
+                dealloc(ptr, layout)
+            };
```

Comment:
Could we use `slice::from_raw_parts()` and `buf.copy_from_slice()` instead of memcpy?

Reply:
This is weird, I'd swear I tried this approach when I started working on this (before Xmas) and hit some issue, but I'm now reviewing the code and can't figure out what the problem could have been.

I'm going to give it another try. Let's see how it works.

Reply:
I've updated the patch series and it works nicely, not sure what kind of issue I hit last time. I also switched the aligned cases to use `self.file.[read|write]`, saving a couple `unsafe` blocks.

I'm still using `pread/pwrite`, as it saves a `lseek` call, but once `read_at/write_at` is stabilized we can cut down another three `unsafe` blocks.